File: src/main/java/dev/vrba/secrethitler/engine/state/transitions/VoteMessage.java
package dev.vrba.secrethitler.engine.state.transitions;

import dev.vrba.secrethitler.engine.Phase;
import dev.vrba.secrethitler.engine.election.Election;
import dev.vrba.secrethitler.engine.election.Vote;
import dev.vrba.secrethitler.engine.state.GameState;
import org.jetbrains.annotations.NotNull;

import java.util.Objects;
import java.util.UUID;
import java.util.function.Supplier;

public class VoteMessage extends GameStateMessage {

    private final Vote vote;

    protected VoteMessage(@NotNull UUID sender, @NotNull Vote vote) {
        super(sender);
        this.vote = vote;
    }

    @Override
    protected boolean validateState(@NotNull GameState state) {
        return state.getPhase() == Phase.VOTING_FOR_THE_GOVERNMENT &&
               state.getElection() != null;
    }

    @Override
    public @NotNull GameState apply(@NotNull GameState state) {
        Election election = Objects.requireNonNull(state.getElection()).vote(this.sender, this.vote);

        if (election.isFinished()) {
            final Supplier<StateTransition> transition = election.isSuccessful()
                ? HandleSuccessfulElection::new
                : HandleFailedElection::new;

            return state.apply(transition);
        }

        return state.withElection(election);
    }
}

File: src/main/kotlin/dev/vrba/secrethitler/engine/state/transitions/VoteMessage.kt
package dev.vrba.secrethitler.engine.state.transitions

import dev.vrba.secrethitler.engine.GamePhase
import dev.vrba.secrethitler.engine.election.Vote
import dev.vrba.secrethitler.engine.state.GameState
import java.util.*

class VoteMessage(sender: UUID, private val vote: Vote) : TransitionMessage(sender) {

    override fun validateState(state: GameState): Boolean =
        state.phase == GamePhase.VotingForTheGovernment &&
                state.election != null

    override fun evaluate(state: GameState): GameState {
        // This is just for smart casting
        if (state.election == null) {
            return state
        }

        val election = state.election.vote(sender, vote)

        if (election.finished) {
            return state.transition(
                if (election.successful) HandleSuccessfulElection()
                else HandleFailedElection()
            )
        }

        return state.copy(election = election)
    }
}